[DNM][RFC] irq: introduce new `irq_set_affinity()` API

ycsin opened the pull request:
> [!NOTE]
> Alternate implementation that doesn't require stubbing every architectures - #77828

`irq_set_affinity(irq, mask)` allows the runtime affinity configuration of an interrupt line, which can be helpful in tuning the performance in a SMP setup (i.e. `irqbalance`).

This commit currently only creates an empty implementation for all of the architectures.

~~I will upstream an implementation for RISCV PLIC in this PR if this API is acceptable,~~ (added an implementation for RISCV PLIC) looking for comments

___

# Testing on `qemu_riscv64_smp`


```
west build -b qemu_riscv64_smp -p always -t run zephyr/samples/hello_world -- \
  -DCONFIG_SHELL=y \
  -DCONFIG_MP_MAX_NUM_CPUS=4 \
  -DCONFIG_PLIC_SHELL=y
```

```
[125/126] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: riscv64
*** Booting Zephyr OS build v3.7.0-1894-g2555dd2bb4f6 ***
Hello World! qemu_riscv64/qemu_virt_riscv64/smp

uart:~$ plic stats get interrupt-controller@c000000 
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         105           0           0           0         123      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         211           0           0           0         229      0x800058fa(0x8000a130)

uart:~$ plic affinity set interrupt-controller@c000000 10 0x2
IRQ 10 affinity set to 0x2
uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         110           0           0         490      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         208           0           0         589      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         315           0           0         695      0x800058fa(0x8000a130)

uart:~$ plic affinity set interrupt-controller@c000000 10 0x4
IRQ 10 affinity set to 0x4
uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         100           0         880      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         206           0         986      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         305           0        1085      0x800058fa(0x8000a130)

uart:~$ plic affinity set interrupt-controller@c000000 10 0x8
IRQ 10 affinity set to 0x8
uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         398         114        1287      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         398         216        1390      0x800058fa(0x8000a130)

uart:~$ plic affinity set interrupt-controller@c000000 10 0xf
IRQ 10 affinity set to 0xF
uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
     0          12          16          14          20          71      0x800045f4(0)
    10         391         432         421         345        1598      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
     0          38          38          40          40         162      0x800045f4(0)
    10         417         457         452         371        1706      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
     0          62          60          65          58         252      0x800045f4(0)
    10         438         484         486         398        1815      0x800058fa(0x8000a130)

uart:~$ 
```

ycsin committed:
irq: introduce new `irq_set_affinity()` API

`irq_set_affinity(irq, mask)` allows the runtime affinity
configuration of an interrupt line, which can be helpful in
tuning the performance in a SMP setup (i.e. `irqbalance`).

This commit creates an empty implementation for all of the
in-tree architectures.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

con-pax commented on include/zephyr/irq.h:
```diff
@@ -309,6 +309,16 @@ void z_smp_global_unlock(unsigned int key);
+/**
+ * @brief Set IRQ affinity.
+ *
+ * This routine configure set of CPU cores that can service the @a irq.
```
nit: "s/This routine configure set/This routine configures a set/"

cfriedt commented:
Looks reasonable to me

fkokosinski commented:
Looks good, but I'd love to see this API "in action" w/ PLIC

ycsin committed:
devicetree: add `DT_INST_NUM_IRQS()`

Add `DT_INST_NUM_IRQS()` to get the number of interrupt lines
of the current `DT_DRV_COMPAT`

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

cfriedt commented on arch/riscv/core/smp.c:
```diff
@@ -74,7 +74,11 @@ void arch_secondary_cpu_init(int hartid)
-#endif
+#ifdef CONFIG_RISCV_PLIC_HAS_IRQ_AFFINITY
+	/* Enable on secondary cores so that they can respond to PLIC */
+	irq_enable(RISCV_IRQ_MEXT);
```
It's somewhat surprising that this has been missing for so long 😅

I guess it would be a non-trivial improvement?

ycsin replied:
> I guess it would mean a non-trivial improvement?

That's yet to be found out, although every core can handle the IRQ after this patch, only 1 core would be able to actually claim it while the others simply waste CPU cycles. 

I guess we will need a new benchmark to test the overall interrupt handling performance

cfriedt replied:
I would say a couple of key things to verify:

1. If an interrupt is pinned only to certain harts / cores, it would be ideal if other cores were simply not interrupted at all.

2. If an interrupt is masked / available to multiple cores, and if the arch must interrupt more than a single core, then the cores that did not receive the interrupt should more or less immediately bounce directly out of the ISR. Maybe some intc's are capable of only selecting / interrupting a single core (that would be more ideal than contention).

Probably would be good to craft a testsuite.

Also perf analyses before and after the change to see if a) there is significant cpu time lost, and b) to measure interrupt latency in some kind of interrupt storm

cfriedt commented on include/zephyr/irq.h:
```diff
@@ -309,6 +309,16 @@ void z_smp_global_unlock(unsigned int key);
+ * @param mask CPU bit mask.
+ */
+#define irq_set_affinity(irq, mask) arch_irq_set_affinity(irq, mask)
+
```
Just a comment really, as I'm not sure what the convention is here, but would it be smart to create `CONFIG_ARCH_HAS_IRQ_AFFINITY`? Non-user-configurable, of course.

As an alternative to a Kconfig option, technically, it could even be on a per-cpu basis if it were a devicetree property (which might help for heterogeneous systems like the big+small combinations present on many ARM SoCs).

The primary reason I think that would be helpful is for testing; the testsuite could simply use that property as a filter, and then setting IRQ affinity could be very easily testable.

The secondary reason is that using a conditional declaration could alleviate the need to implement stubs everywhere (although maybe that is the convention).

What do you think?

ycsin replied:
Thanks for this very good feedback!

> Just a comment really, as I'm not sure what the convention is here, but would it be smart to create `CONFIG_ARCH_HAS_IRQ_AFFINITY`? Non-user-configurable, of course.

I thought about doing something like this initially:

```c
#ifdef CONFIG_ARCH_HAS_IRQ_AFFINITY
#define irq_set_affinity(irq, mask) arch_irq_set_affinity(irq, mask)
#else
#define irq_set_affinity(irq, mask)
#endif /* CONFIG_ARCH_HAS_IRQ_AFFINITY */
```

but I didn't find any example of unimplemented IRQ arch APIs so I went with the current implementation (empty stub)

> The primary reason I think that would be helpful is for testing; the testsuite could simply use that property as a filter, and then setting IRQ affinity could be very easily testable.

This is a very compelling one

> As an alternative to a Kconfig option, technically, it could even be on a per-cpu basis if it were a devicetree property (which might help for heterogeneous systems like the big+small combinations present on many ARM SoCs).

Any pointers/examples?

cfriedt replied:
Let's see what others think. @npitre,  @carlocaione, @ceolin, @dcpleung 

It's possible that some architectures might not support IRQ pinning / affinity, but I also appreciate having a consistent arch interface.

dcpleung replied:
I guess as long as if can be disabled, this should be fine.

dcpleung replied:
Well... one thing to think about is that sometimes `arch_irq_*` are being implemented in the interrupt controller drivers. So in the arch layer, maybe follow the convention of `z_<arch>_irq_*` and the actual `arch_irq_*` in whichever files they need to be?

ycsin replied:
> So in the arch layer, maybe follow the convention of z_<arch>_irq_* and the actual arch_irq_* in whichever files they need to be?

Not sure if I understood you correctly - this PR currently declare empty `arch_irq_set_affinity()` stubs according to the convention of the way each arch implements the others existing IRQ APIs, so that they can "fill in the blanks" later if they chose to implement this API, otherwise calling this API would just do nothing.

Anything further than that seems to highly arch-specific, where some of the arch simply redirects to another function (i.e. `arm_gic_irq_is_enabled()`), while some arch implemented it in place (i.e. `arch/mips/core/irq_manage.c`) or in an interrupt controller driver (i.e. `intc_swerv_pic.c`). Making these APIs consistent is probably out of the scope of this PR.

The other approach at #77828 will create an arch-level Kconfig, which will compile the `irq_set_affinity()` to nothing if it isn't implemented/enabled.

dcpleung replied:
If `arch_irq_set_affinity()` is implemented in the interrupt driver, they can then call the architecture `z_<arch>_irq_set_affinity()` (for example) if needed.

Though, ignoring the need for `z_<arch>_irq_*`, the introduction of `arch_irq_set_affinity()` needs to accommodate different implementation styles where it can be implemented in the interrupt drivers and not just inside the architecture layer.

ycsin replied:
> If `arch_irq_set_affinity()` is implemented in the interrupt driver, they can then call the architecture `z_<arch>_irq_set_affinity()` (for example) if needed.

I don't get this - why would they want to call the architecture `z_<arch>_irq_set_affinity()`, if `arch_irq_set_affinity()` is implemented in the driver?

> the introduction of `arch_irq_set_affinity()` needs to accommodate different implementation styles where it can be implemented in the interrupt drivers and not just inside the architecture layer.

Do you mean we should allow `arch_irq_set_affinity()` to be implemented/overridden anywhere? I guess this is basically #77828?

dcpleung replied:
The `z_<arch>_irq_*` is simply where some architectures decide to do IRQ managements at arch level code. Setting affinity may or may not be needed at arch code.

#77828 would definitely be a cleaner version, and would allow driver to implement the function to set affinity.

ycsin committed:
drivers: intc: plic: refactor lock in `plic_irq_enable_set_state()`

Move the lock out from the `plic_irq_enable_set_state()` function
to cover the entire configuration process, so the whole of
enable/disable is atomic.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: intc: plic: implement `irq_set_affinity()`

- Implement `irq_set_affinity()` in RISCV PLIC.
- Added new affinity shell command to get/set the irq(s)
  affinity in runtime, when `0` is sent as the `local_irq`, it
  means set/get all IRQs affinity.
- Some minor optimizations

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

cfriedt commented on drivers/interrupt_controller/Kconfig.plic:
```diff
@@ -13,11 +13,85 @@ config PLIC
+choice PLIC_IRQ_COUNT_TYPE
+	prompt "Variable type used to store IRQ hit count"
+	depends on PLIC_SHELL_IRQ_COUNT
+	default PLIC_IRQ_COUNT_UINT16
+
+config PLIC_IRQ_COUNT_UINT8
+	bool "Stores up to 255"
+	help
+	  Use uint8_t as the underlying type.
+	  This stores up to 255 hits.
+
+config PLIC_IRQ_COUNT_UINT16
+	bool "Stores up to 65535"
+	help
+	  Use uint16_t as the underlying type.
+	  This stores up to 65535 hits.
+
+config PLIC_IRQ_COUNT_UINT32
+	bool "Stores up to 4294967295"
+	help
+	  Use uint32_t as the underlying type.
+	  This stores up to 4294967295 hits.
+
+config PLIC_IRQ_COUNT_UINT64
+	bool "Stores up to 18446744073709552000"
+	help
+	  Use uint64_t as the underlying type.
+	  This stores up to 18446744073709552000 hits.
+
+endchoice # PLIC_IRQ_COUNT_TYPE
+
+config PLIC_IRQ_COUNT_WIDTH
+	int
+	default 5 if PLIC_IRQ_COUNT_UINT8 # width(255) is 3, but width(CPU##) is 5
+	default 5 if PLIC_IRQ_COUNT_UINT16
+	default 10 if PLIC_IRQ_COUNT_UINT32
+	default 20 if PLIC_IRQ_COUNT_UINT64
+	depends on PLIC_SHELL_IRQ_COUNT
+	help
+	  Internal helper to get the width for a variable type
```
Is this change necessary within this PR? I kind of worry that it adds a bit of noise, and it's not clear to me what the immediate benefit would be to move away from `uint16_t`.

Do we really anticipate over 65k interrupts to go unserviced? That seems like it would be a deal-breaker in many real-time systems.

ycsin replied:
It stores the number of time an IRQ has fired, not unserviced

cfriedt commented on drivers/interrupt_controller/CMakeLists.txt:
```diff
@@ -50,7 +50,7 @@ endif()
   message(WARNING "
       WARNING:  `CONFIG_PLIC_SHELL` is enabled.
-      This can use quite a bit of RAM (PLICs * IRQs * sizeof(uint16_t))"
+      This can use quite a bit of RAM (PLICs * IRQs * sizeof(PLIC_IRQ_COUNT_TYPE))"
```
Unless there is a significant reason to change from `uint16_t` within this PR, I would say it's simpler to keep the original type.

ycsin replied:
When SMP IPI is done over PLIC, it overflows the `uint16_t` very quickly, so I thought that it might be nice to make it configurable, but I have removed that from this PR to reduce noice

andyross reviewed (commented):
(no complaints, but I like the other stubless version better, personally)

cfriedt commented:
> (no complaints, but I like the other stubless version better, personally)
> 

Same - the alternate has my vote

ycsin commented:
closed for now, as #77828 seems to be the preferred implementation
